cajita/unit_test/tstArray2d.hpp

```diff
@@ -0,0 +1,388 @@
+    auto ghosted_space = array->layout()->indexSpace( Ghost(), Local() );
+    for ( long i = 0; i < ghosted_space.extent( Dim::I ); ++i )
+        for ( long j = 0; j < ghosted_space.extent( Dim::J ); ++j )
+            for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )
+                EXPECT_EQ( host_view( i, j, l ), 2.0 );
+
+    // Scale the entire array with a single value.
+    ArrayOp::scale( *array, 0.5, Ghost() );
+    Kokkos::deep_copy( host_view, array->view() );
+    for ( long i = 0; i < ghosted_space.extent( Dim::I ); ++i )
+        for ( long j = 0; j < ghosted_space.extent( Dim::J ); ++j )
+            for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )
+                EXPECT_EQ( host_view( i, j, l ), 1.0 );
+
+    // Scale each array component by a different value.
+    std::vector<double> scales = { 2.3, 1.5, 8.9, -12.1 };
+    ArrayOp::scale( *array, scales, Ghost() );
+    Kokkos::deep_copy( host_view, array->view() );
+    for ( long i = 0; i < ghosted_space.extent( Dim::I ); ++i )
+        for ( long j = 0; j < ghosted_space.extent( Dim::J ); ++j )
+            for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )
+                EXPECT_EQ( host_view( i, j, l ), scales[l] );
+
+    // Create another array and update.
+    auto array_2 = createArray<double, TEST_DEVICE>( label, cell_layout );
+    ArrayOp::assign( *array_2, 0.5, Ghost() );
+    ArrayOp::update( *array, 3.0, *array_2, 2.0, Ghost() );
+    Kokkos::deep_copy( host_view, array->view() );
+    for ( long i = 0; i < ghosted_space.extent( Dim::I ); ++i )
+        for ( long j = 0; j < ghosted_space.extent( Dim::J ); ++j )
+            for ( long l = 0; l < ghosted_space.extent( 2 ); ++l )
+                EXPECT_EQ( host_view( i, j, l ), 3.0 * scales[l] + 1.0 );
+
+    // Check the subarray.
+    auto subarray = createSubarray( *array, 2, 4 );
+    auto sub_ghosted_space = subarray->layout()->indexSpace( Ghost(), Local() );
+    EXPECT_EQ( sub_ghosted_space.rank(), 3 );
+    EXPECT_EQ( sub_ghosted_space.extent( Dim::I ),
+               ghosted_space.extent( Dim::I ) );
+    EXPECT_EQ( sub_ghosted_space.extent( Dim::J ),
+               ghosted_space.extent( Dim::J ) );
+    auto host_subview = Kokkos::create_mirror_view_and_copy(
+        Kokkos::HostSpace(), subarray->view() );
+    for ( long i = 0; i < sub_ghosted_space.extent( Dim::I ); ++i )
+        for ( long j = 0; j < sub_ghosted_space.extent( Dim::J ); ++j )
+            for ( long l = 0; l < sub_ghosted_space.extent( 2 ); ++l )
+                EXPECT_EQ( host_subview( i, j, l ), 3.0 * scales[l + 2] + 1.0 );
+
+#ifndef KOKKOS_ENABLE_OPENMPTARGET // FIXME_OPENMPTARGET
+    // Compute the dot product of the two arrays.
+    std::vector<double> dots( dofs_per_cell );
+    ArrayOp::dot( *array, *array_2, dots );
+    int total_num_node = global_grid->globalNumEntity( Node(), Dim::I ) *
+                         global_grid->globalNumEntity( Node(), Dim::J );
```

Comment:
it is valid since it is periodic condition. If it is non-periodic, it needs to be Cell()

Reply:
Actually you are right! This should be `Cell`. It is a typo.